### ISSUE TITLE
Error 2 when attempting to build cxx object bar.cpp.o

### ISSUE THREAD

**Issue by AceOfHurts** [jaagr/polybar#644]
I have been trying to build polybar from the source and keep running into this issue.  I have never filed an issue before so I'm not sure how much I should include.  In the terminal before the error 2 stuff there is a long list of errors.  They are all relating to xcb_randr_get_monitors_monitors and .  I attempted to use the fix in the issue thread #633 but still ran into troubles.  

bin/CMakeFiles/polybar.dir/build.make:182: recipe for target 'bin/CMakeFiles/polybar.dir/components/bar.cpp.o' failed
make[2]: *** [bin/CMakeFiles/polybar.dir/components/bar.cpp.o] Error 1
CMakeFiles/Makefile2:798: recipe for target 'bin/CMakeFiles/polybar.dir/all' failed
make[1]: *** [bin/CMakeFiles/polybar.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by patrick96** [comment]
Can you post the output you got when you ran `cmake ..` and also the whole output of make (in a gist or something like that)

**Comment by AceOfHurts** [comment]
Here is the gist for when I run [make](https://gist.github.comAceOfHurtsd19eed23ac79c30a7058991d8ac1fbff).  Here is the gist for running only [cmake ..](https://gist.github.com/AceOfHurts/2a57fa00cd6cc6a13bfa5687eac0ad4f) and lastly when I run [cmake with clang](https://gist.github.com/AceOfHurts/2624d42db9c586da3e86acdecf49b71e).  Unfortunately running cmake with clang didn't seem to work.

**Comment by patrick96** [comment]
Did you run `make` after you configured cmake with clang? If so can you post the output for that too.

It seems this may be a problem with your `libxcb`, what version do you have installed?
Also what distro are you running?

**Comment by AceOfHurts** [comment]
Yes I did run make after I configured cmake with clang but I still got an Error2 here is the [gist](https://gist.github.com/AceOfHurts/b15b64bfa956514b98624e2ee1aae385).  I am running Ubuntu 17.04 with openbox as my wm.  As far as libxcb goes I am running 1.12.

**Comment by patrick96** [comment]
Alright, that is really strange, it complains about a few functions that are not defined, but they should be part of the `libxcb` version you are running.

Can you make sure that you have installed all of the packages listed in the [Wiki](https://github.com/jaagr/polybar/wiki/Compiling)

**Comment by AceOfHurts** [comment]
I checked and I have all of the packages installed.  I had run the command `sudo apt install cmake cmake-data libcairo2-dev libxcb1-dev libxcb-ewmh-dev libxcb-icccm4-dev libxcb-image0-dev libxcb-randr0-dev libxcb-util0-dev libxcb-xkb-dev pkg-config python-xcbgen xcb-proto libxcb-xrm-dev i3-wm libasound2-dev libmpdclient-dev libiw-dev libcurl4-openssl-dev` and everything said it was up to date.

**Comment by patrick96** [comment]
Maybe the `randr.h` on your system is missing some function.
Can you give me the output of the following two commands:
`dpkg -S /usr/include/xcb/randr.h`
`dpkg -S $(man -w xcb_randr_get_monitors_reply)`

**Comment by AceOfHurts** [comment]
When I run `dpkg -S /usr/include/xcb/randr.h` it outputs libxcb-randr0-dev:amd64: /usr/include/xcb/randr.h

When `dpkg -S $(man -w xcb_randr_get_monitors_reply)` is run it outputs [this.](https://gist.github.com/AceOfHurts/ed610b5b797922cb825187324c25b31b)

**Comment by patrick96** [comment]
Man this was a difficult bug! For future reference, I'll describe exactly what happened, at the end there is a section on how to actually fix it.

I had to install Ubuntu on a VM to actually see what the problem was:

![2017-07-12-235941_243x278_scrot](https://user-images.githubusercontent.com/2452038/28142158-eb04b028-675f-11e7-9109-3f438465a6ab.png) ![2017-07-13-001331_279x260_scrot](https://user-images.githubusercontent.com/2452038/28142203-1e245652-6760-11e7-9d38-9952d3fc13ca.png)

On the left is the output of `cmake` on my system and on the right is the output on the Ubuntu VM. Notice how `xcb-randr (monitor support)` is greyed out in the VM. This is something you cannot see from the logs.

Digging deeper that extension is checked [here](https://github.com/jaagr/polybar/blob/2314ca030c56e9d1951be5f9b707b638b38a649c/cmake/02-opts.cmake#L11):
```cmake
checklib(WITH_XRANDR_MONITORS "pkg-config" "xcb-randr>=1.12")
```
So the difference between my system and the VM seems to be the `xcb-randr` version. On my system it's `1.12` and it's `1.11` in the VM.
Polybar actually uses that `WITH_XRANDR_MONITORS` variable to disable calls to the functions that are new in the `1.12` version (see [src/x11/extensions/randr.cpp:88](https://github.com/jaagr/polybar/blob/2314ca030c56e9d1951be5f9b707b638b38a649c/src/x11/extensions/randr.cpp#L88)).
From the log I could see that the unknown function was called from a file in the `xpp` library, so I thought I could also use that variable to remove ~those~ function calls.

But the file that complained about the `undeclared identifier 'xcb_randr_get_monitors_reply'` is a file that does not actually exist in the `xpp` repository:
```
/home/maslin/polybar/lib/xpp/include/xpp/proto/randr.hpp:3026:42: error: use of undeclared identifier 'xcb_randr_get_monitors_reply'; did you mean 'xcb_randr_get_providers_reply'?
```
All the files in the `proto` folder are generated from the xml files provided by `xcb-proto`, now if you check the version of `xcb-proto` on your system it will probably be `1.12-1` while the version of `libxcb-randr0-dev` is `1.11-1`. This is the problem,
 the header files generated from the `xcb-proto` xml files rely on `<xcb/randr.h>` from `libxcb-randr0-dev`.
And because the versions of those two packages don't match the `randr.h` header file does not include all the functions that `xcb-proto` assumes are there.
So we'll have to make the versions match either by upgrading all the `libxcb-*` packages to `1.12-1` or downgrading `xcb-proto` to `1.11-1`. I haven't found any `libxcb-*` ubuntu packages for version `1.12-1` so I downgraded `xcb-proto` (see last section)


**TL;DR Just tell me how to compile it!**

You will need to downgrade to `xcb-proto` to version `1.11-1`, I didn't have any luck with using `apt` to downgrade, so I grabbed the source files from [launchpad](https://launchpad.net/ubuntu/+source/xcb-proto/1.11-1) and compiled from source.
You have to get both `xcb-proto_1.11.orig.tar.gz` and `xcb-proto_1.11-1.diff.gz`.

Now remove the already installed `xcb-proto` package and run the following command inside the folder where you downloaded the files to. This will compile `xcb-proto-1.11-1` and install it on your system.
~~~bash
tar -xzvf xcb-proto_1.11.orig.tar.gz
gzip -d xcb-proto_1.11-1.diff.gz
cd xcb-proto-1.11
patch -p1 <../xcb-proto_1.11-1.diff
./configure
make
sudo make install
~~~
Note that it will not be installed via `apt` and thus to uninstall you'll need to run `sudo make uninstall` from within the `xcb-proto-1.11` folder.

Now you should be able to compile polybar without problems, just clear the build directory first.

**Comment by AceOfHurts** [comment]
That worked up until it got to renderer.cpp.o.  From looking at the error it seems like I might need a different version of Cairo-fc.  Here is the [output.](https://gist.github.com/AceOfHurts/1e27aff8249810c2ef10ecf1e02a9e03) Tried building it with both cmake .. and cmake using clang.

**Comment by patrick96** [comment]
What's the output of `dpkg -S /usr/include/cairo/cairo-features.h` and can you post the contents of `/usr/include/cairo/cairo-features.h`

**Comment by AceOfHurts** [comment]
Output of dpkg -S /usr/include/cairo/cairo-features.h `libcairo2-dev: /usr/include/cairo/cairo-features.h`
Contents of [cairo-features.h ](https://gist.github.com/AceOfHurts/3f69d9de5db902df48de545e02191486)

**Comment by patrick96** [comment]
And also `dpkg -s libcairo2-dev` as well as the contents of `/usr/include/cairo/cairo.h`

**Comment by AceOfHurts** [comment]
Here is the gist for [libcario2-dev](https://gist.github.com/AceOfHurts/d314c6c173fb896cb0b84ee45ea56f29) and then here is the gist for [cairo.h](https://gist.github.com/AceOfHurts/8ddfe1ecb86279f27b4792587cbd0c30)

**Comment by patrick96** [comment]
That's the same as on my VM. There probably is some cairo dependency that's not listed on the wiki. I'll need to check that when I get home

**Comment by patrick96** [comment]
I am not able to reproduce this on a clean ubutnu 17.04 VM.
Did you compile `libcairo2-dev` yourself?

**Comment by AceOfHurts** [comment]
I did but was missing a few things.  I'll uninstall libcario2 and then delete whats left.

**Comment by patrick96** [comment]
Glad to hear it works.
